federation/pkg/federation-controller/service/dns.go

```diff
@@ -124,11 +121,34 @@ func getRrset(dnsName string, rrsetsInterface dnsprovider.ResourceRecordSets) (d
 	return returnVal, nil
 }
 
+/* getResolvedEndpoints perfoms DNS resolution on the provided slice of endpoints (which might be DNS names or IPv4 addresses)
+   and returns a list of IPv4 addresses.  If any of the endpoints are neither valid IPv4 addresses nor resolvable DNS names,
+   non-nil error is also returned (possibly along with a partially complete list of resolved endpoints.
+*/
+func getResolvedEndpoints(endpoints []string) ([]string, error) {
+	resolvedEndpoints := make([]string, len(endpoints))
+	for _, endpoint := range endpoints {
+		if net.ParseIP(endpoint) == nil {
+			// It's not a valid IP address, so assume it's a DNS name, and try to resolve it,
+			// replacing it's DNS name with it's IP addresses in expandedEndpoints
+			ipAddrs, err := net.LookupHost(endpoint)
+			if err != nil {
+				return resolvedEndpoints, err
+			}
+			resolvedEndpoints = append(resolvedEndpoints, ipAddrs...)
```

Comment:
It looks like this has the same problem what we had during the Ubernetes hackathon. 

```Go
resolvedEndpoints := make([]string, len(endpoints))
```

Creates a slice of length len(endpoints). And then you append after that. So effectively you have a slice of `2 * len(endpoints)`

Sorry for not catching this earlier.